### ISSUE TITLE
Mi Band 1A recognized as 1S during firmware update

### ISSUE THREAD

**Issue by Vzaa** [Freeyourgadget/Gadgetbridge#364]
I recently bought a Mi Band 1A and it came with firmware version 5.15.11.19 installed according to Gadgetbridge 0.11.2. Wiki recommends firmware version 5.15.7.14 so I tried installing the matching Mili_1a.fw and it crashes Gadgetbridge every time I try it. I looked around the issues and the wiki but couldn't find a solution.

Checking the logs it seems to me like my 1A was detected as a 1S:

`07-26 19:58:42.993 10508 10508 E nodomain.freeyourgadget.gadgetbridge.LoggingExceptionHandler: Caused by: java.lang.IllegalArgumentException: preparing single fw not allowed for 1S
`

I think it might be caused by wrong detection of heart rate support:

https://github.com/Freeyourgadget/Gadgetbridge/blob/0.11.2/app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/miband/operations/UpdateFirmwareOperation.java#L53

**Comment by cpfeiffer** [comment]
Wow, I'm glad we have that check there. Better abort the app than crash and burn the band...

Can you share some more bits of the logfile? It should contain some device information which is used to determine the correct firmware. Look for a line containing `Device info:` or search for `hwVersion`.

**Comment by Vzaa** [comment]
I've reproduced the crash with logging enabled. I've attached the full log I got from launch to crash. I believe these are the related lines:

`04:44:22.160 [Binder_1] WARN  n.f.g.s.d.m.MiBandSupport - Device info: DeviceInfo{deviceId='C872BB5F05000882', profileVersion=33555968, fwVersion=84871955, hwVersion=8, feature=5, appearance=0, fw2Version (hr)=-1}
`

`04:44:45.702 [Binder_1] WARN  n.f.g.s.d.m.MiBandSupport - Device info: DeviceInfo{deviceId='C872BB5F05000882', profileVersion=33555968, fwVersion=84871955, hwVersion=8, feature=5, appearance=0, fw2Version (hr)=-1}
`


[gadgetbridge.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/385141/gadgetbridge.txt)

**Comment by cpfeiffer** [comment]
Thanks a lot for the log. We need to rethink the way we recognize the different Mi devices.

**Comment by frankycan** [comment]
Hi! I have a similar error. My MiBand is the Model 1, the one with coloured-LEDs. The current FW is 1.0.11.7. I have tried to upgrade with some FW files, with the ones listed at the Readme.txt and other ones that are not listed, and I'm unable to upgrade/downgrade the FW. I always get an error like "This firmware is not compatible with your device" (not exactly that String).

The smartphone where I'm trying to upgrade is a Motorola Nexus 6 with Android 7.0.

I attach the log generated at the failed upgrade process, but this lines makes me think that is a bad detection with the bands. The FW file used at the log is the 1.0.12.

[gadgetbridge.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/810855/gadgetbridge.txt)

PS: At the log appears a 'Mi Band 2' from a workmate, but it's not mine

**Comment by cpfeiffer** [comment]
The log file is a little weird. It mostly displays output from the discovery, and suddenly some firmware detection stuff. But not connection at all.

Can you
- connect to the device and then
- try to open the *.fw file

and provide a log of that?

**Comment by frankycan** [comment]
Thanks for the answer! 

I connected the MiBand and then tried to update the FW and this is the result log. Hope this log will be more useful than the other one :)

[gadgetbridge.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/818798/gadgetbridge.txt)

**Comment by cpfeiffer** [comment]
@frankycan Please try 0.18.4 and see if this improves your situation. Your band reports a very different hardware version (8) which we did not recognize as Mi Band 1.

**Comment by frankycan** [comment]
Awesome! I downloaded the last FDroid APK but it doesn't worked (version 0.18.0), but when I executed the GitHub source code into my smartphone (version 0.18.3) it updated the firmware at first attempt!!

The fw upgrade was from 1.0.11.7 to 1.0.12.  Later I upgraded from 1.0.12 to 1.0.12.5

Unfortunately, MiFit app keeps trying to update the MiBand fw although it has the latest one (extracted from the APK). It fails every time the MiFit app tries to update the MiBand fw.

Thank you!!
